client/src/client.rs

```diff
@@ -452,56 +452,79 @@ impl Client {
-
-                self.submit_transaction(transaction)?;
+        std::thread::scope(|spawner| {
+            let submitter_handle = spawner.spawn(move || -> Result<()> {
+                // Do not submit transaction if event listener is failed to initialize
+                if init_receiver
+                    .blocking_recv()
+                    .wrap_err("Failed to receive init message.")?
+                {
+                    self.submit_transaction(transaction)?;
+                }
                 Ok(())
             });
 
-            let confirmation_res = client.listen_for_tx_confirmation(hash, &init_sender);
+            let confirmation_res = self.listen_for_tx_confirmation(init_sender, hash);
 
             match submitter_handle.join() {
                 Ok(Ok(())) => confirmation_res,
                 Ok(Err(e)) => Err(e).wrap_err("Transaction submitter thread exited with error"),
                 Err(_) => Err(eyre!("Transaction submitter thread panicked")),
             }
-        });
-
-        match scope_res {
-            Ok(res) => res,
-            Err(_) => Err(eyre!("Transaction submitter thread panicked")),
-        }
+        })
     }
 
-    // TODO: introduce timeout
     fn listen_for_tx_confirmation(
         &self,
+        init_sender: tokio::sync::oneshot::Sender<bool>,
         hash: HashOf<SignedTransaction>,
-        init_sender: &mpsc::Sender<()>,
     ) -> Result<HashOf<VersionedSignedTransaction>> {
-        let event_iterator = self
-            .listen_for_events(PipelineEventFilter::new().hash(hash.into()).into())
-            .wrap_err("Failed to establish event listener connection")?;
+        let deadline = tokio::time::Instant::now() + self.transaction_status_timeout;
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()?;
+
+        rt.block_on(async {
+            let mut event_iterator = {
+                let event_iterator_result = tokio::time::timeout_at(
+                    deadline,
+                    self.listen_for_events_async(
+                        PipelineEventFilter::new().hash(hash.into()).into(),
+                    ),
+                )
+                .await
+                .map_err(Into::into)
+                .and_then(std::convert::identity)
+                .wrap_err("Failed to establish event listener connection");
+                let _send_result = init_sender.send(event_iterator_result.is_ok());
+                event_iterator_result?
+            };
 
-        init_sender
-            .send(())
-            .wrap_err("Failed to send init message through init channel")?;
+            let result = tokio::time::timeout_at(
+                deadline,
+                Self::listen_for_tx_confirmation_loop(&mut event_iterator, hash),
+            )
+            .await
+            .map_err(Into::into)
+            .and_then(std::convert::identity);
```

Comment:
Odd construction.

Reply:
Intention here is to flatten `Result`.
1. initially we have `Result<Result<T, Report>, E>`
2. by `.map_err(Into::into)` we convert `E` into `Report`
3. by `.and_then(std::convert::identity)` we flatten result from `Result<Result<T, Report>, Report>` into `Result<T, Report>` (i take inspiration from nightly api [Result::flatten](https://doc.rust-lang.org/src/core/result.rs.html#1769))